File: server/protocol_smoke.py
#!/usr/bin/env python3

import argparse
import http.client
import json
import socket
import ssl
import time
import uuid
from pathlib import Path
from typing import Any, Optional
from urllib.parse import urlparse

from server.gateway_server import (
    encrypt_payload,
    expected_signature,
    load_devices,
)


class SniHttpsConnection(http.client.HTTPSConnection):
    """Connect to one host while validating TLS for another server name."""

    def __init__(
        self,
        connect_host: str,
        server_name: str,
        port: int,
        context: ssl.SSLContext,
        timeout: float,
    ) -> None:
        super().__init__(
            connect_host,
            port=port,
            context=context,
            timeout=timeout,
        )
        self._server_name = server_name

    def connect(self) -> None:
        sock = socket.create_connection(
            (self.host, self.port),
            self.timeout,
            self.source_address,
        )
        if self._tunnel_host:
            self.sock = sock
            self._tunnel()
            sock = self.sock
        self.sock = self._context.wrap_socket(
            sock,
            server_hostname=self._server_name,
        )


def build_smoke_request(
    device_id: str,
    secret: bytes,
    now_ms: Optional[int] = None,
) -> tuple[bytes, dict[str, str]]:
    timestamp_ms = now_ms if now_ms is not None else int(time.time() * 1000)
    unique = uuid.uuid4().hex
    envelope: dict[str, Any] = {
        "schemaVersion": 1,
        "deliveryId": f"production-smoke-{unique}",
        "sourceEventId": f"production-smoke-{unique}",
        "eventType": "LOCAL_SELF_TEST",
        "createdAt": timestamp_ms,
        "subscriptionId": None,
        "slotIndex": None,
        "payload": {
            "type": "LOCAL_SELF_TEST",
            "source": "production-deployment-check",
        },
    }
    encrypted = encrypt_payload(envelope, device_id, secret)
    body = json.dumps(
        encrypted,
        separators=(",", ":"),
        ensure_ascii=False,
    ).encode("utf-8")
    nonce = str(uuid.uuid4())
    idempotency_key = f"production-smoke-{unique}"
    return body, {
        "Content-Type": "application/json; charset=utf-8",
        "X-Gateway-Device": device_id,
        "X-Gateway-Timestamp": str(timestamp_ms),
        "X-Gateway-Nonce": nonce,
        "X-Gateway-Signature": expected_signature(
            secret,
            timestamp_ms,
            nonce,
            device_id,
            idempotency_key,
            body,
        ),
        "Idempotency-Key": idempotency_key,
    }


def run_smoke(
    endpoint: str,
    config_path: Path,
    device_id: Optional[str] = None,
    connect_host: Optional[str] = None,
    ca_file: Optional[Path] = None,
) -> int:
    parsed = urlparse(endpoint)
    if (
        parsed.scheme != "https"
        or not parsed.hostname
        or parsed.username
        or parsed.password
        or parsed.query
        or parsed.fragment
        or parsed.path not in ("", "/")
    ):
        raise ValueError("endpoint must be an HTTPS origin URL")
    devices = load_devices(config_path)
    selected_device = device_id or next(iter(devices))
    secret = devices.get(selected_device)
    if secret is None:
        raise ValueError("selected device is not configured")

    body, headers = build_smoke_request(selected_device, secret)
    context = ssl.create_default_context(
        cafile=str(ca_file) if ca_file is not None else None
    )
    port = parsed.port or 443
    connection = SniHttpsConnection(
        connect_host or parsed.hostname,
        parsed.hostname,
        port,
        context,
        timeout=10,
    )
    host_header = parsed.hostname
    if parsed.port is not None and parsed.port != 443:
        host_header = f"{host_header}:{parsed.port}"
    headers["Host"] = host_header
    try:
        connection.request("POST", "/v1/events", body=body, headers=headers)
        response = connection.getresponse()
        response_body = response.read(16_384)
    finally:
        connection.close()
    if response.status not in (200, 201):
        raise RuntimeError(f"protocol smoke failed with HTTP {response.status}")
    try:
        value = json.loads(response_body)
    except (UnicodeDecodeError, json.JSONDecodeError) as error:
        raise RuntimeError("protocol smoke returned invalid JSON") from error
    if not isinstance(value, dict) or value.get("accepted") is not True:
        raise RuntimeError("protocol smoke was not accepted")
    return response.status


def main() -> None:
    parser = argparse.ArgumentParser(
        description="Send one signed encrypted event through the production proxy."
    )
    parser.add_argument("--url", required=True)
    parser.add_argument("--config", type=Path, required=True)
    parser.add_argument("--device-id")
    parser.add_argument(
        "--connect-host",
        help="TCP destination override; TLS SNI and Host still use --url.",
    )
    parser.add_argument("--ca-file", type=Path)
    args = parser.parse_args()
    status = run_smoke(
        args.url,
        args.config,
        device_id=args.device_id,
        connect_host=args.connect_host,
        ca_file=args.ca_file,
    )
    print(f"Signed encrypted protocol smoke test: PASS (HTTP {status})")


if __name__ == "__main__":
    main()

File: server-node/src/operations/protocol-smoke.ts
import { request } from 'node:https';
import { basename } from 'node:path';
import { randomUUID } from 'node:crypto';
import { readFileSync } from 'node:fs';
import { loadRuntimeConfig } from '../config/runtime-config.js';
import { encryptPayload, type Envelope } from '../crypto/payload-crypto.js';
import { expectedSignature } from '../crypto/request-signature.js';

export async function runProtocolSmoke(options: {
  url: string;
  configPath: string;
  deviceId?: string;
  connectHost?: string;
  caFile?: string;
}): Promise<number> {
  const endpoint = new URL(options.url);
  if (
    endpoint.protocol !== 'https:'
    || !endpoint.hostname
    || endpoint.username
    || endpoint.password
    || endpoint.search
    || endpoint.hash
    || (endpoint.pathname !== '/' && endpoint.pathname !== '')
  ) {
    throw new Error('endpoint must be an HTTPS origin URL');
  }

  const config = loadRuntimeConfig(options.configPath);
  const deviceId = options.deviceId ?? config.configuredDevices.keys().next().value;
  const secret = deviceId ? config.configuredDevices.get(deviceId) : undefined;
  if (!deviceId || !secret) throw new Error('selected device is not configured');

  const timestamp = Date.now();
  const unique = randomUUID().replaceAll('-', '');
  const envelope: Envelope = {
    schemaVersion: 1,
    deliveryId: `production-smoke-${unique}`,
    sourceEventId: `production-smoke-${unique}`,
    eventType: 'LOCAL_SELF_TEST',
    createdAt: timestamp,
    subscriptionId: null,
    slotIndex: null,
    payload: { type: 'LOCAL_SELF_TEST', source: 'production-deployment-check' },
  };
  const body = Buffer.from(JSON.stringify(encryptPayload(envelope, deviceId, secret)));
  const nonce = randomUUID();
  const idempotencyKey = `production-smoke-${unique}`;
  const signature = expectedSignature(secret, timestamp, nonce, deviceId, idempotencyKey, body);

  const response = await sendRequest(endpoint, body, {
    deviceId,
    timestamp,
    nonce,
    idempotencyKey,
    signature,
    connectHost: options.connectHost,
    caFile: options.caFile,
  });
  if (response.status !== 200 && response.status !== 201) {
    throw new Error(`protocol smoke failed with HTTP ${response.status}`);
  }
  let value: unknown;
  try {
    value = JSON.parse(response.body);
  } catch {
    throw new Error('protocol smoke returned invalid JSON');
  }
  if (typeof value !== 'object' || value === null || (value as Record<string, unknown>).accepted !== true) {
    throw new Error('protocol smoke was not accepted');
  }
  return response.status;
}

function sendRequest(
  endpoint: URL,
  body: Buffer,
  options: {
    deviceId: string;
    timestamp: number;
    nonce: string;
    idempotencyKey: string;
    signature: string;
    connectHost?: string;
    caFile?: string;
  },
): Promise<{ status: number; body: string }> {
  return new Promise((resolve, reject) => {
    const req = request({
      protocol: 'https:',
      hostname: options.connectHost ?? endpoint.hostname,
      port: endpoint.port ? Number(endpoint.port) : 443,
      path: '/v1/events',
      method: 'POST',
      servername: endpoint.hostname,
      ca: options.caFile ? readFileSync(options.caFile) : undefined,
      headers: {
        Host: endpoint.host,
        'Content-Type': 'application/json; charset=utf-8',
        'Content-Length': String(body.length),
        'X-Gateway-Device': options.deviceId,
        'X-Gateway-Timestamp': String(options.timestamp),
        'X-Gateway-Nonce': options.nonce,
        'X-Gateway-Signature': options.signature,
        'Idempotency-Key': options.idempotencyKey,
      },
      timeout: 10_000,
    }, response => {
      const chunks: Buffer[] = [];
      response.on('data', chunk => {
        if (chunks.reduce((total, value) => total + value.length, 0) < 16_384) {
          chunks.push(Buffer.from(chunk));
        }
      });
      response.on('end', () => resolve({
        status: response.statusCode ?? 0,
        body: Buffer.concat(chunks).toString('utf-8'),
      }));
    });
    req.on('timeout', () => req.destroy(new Error('protocol smoke timed out')));
    req.on('error', reject);
    req.end(body);
  });
}

async function main(): Promise<void> {
  const args = parseArguments(process.argv.slice(2));
  const status = await runProtocolSmoke(args);
  console.log(`Signed encrypted protocol smoke test: PASS (HTTP ${status})`);
}

function parseArguments(argv: string[]): {
  url: string;
  configPath: string;
  deviceId?: string;
  connectHost?: string;
  caFile?: string;
} {
  const values = new Map<string, string>();
  for (let index = 0; index < argv.length; index += 2) {
    const key = argv[index];
    const value = argv[index + 1];
    if (!key?.startsWith('--') || value === undefined) throw new Error('invalid arguments');
    values.set(key, value);
  }
  const url = values.get('--url');
  const configPath = values.get('--config');
  if (!url || !configPath) throw new Error('--url and --config are required');
  return {
    url,
    configPath,
    deviceId: values.get('--device-id'),
    connectHost: values.get('--connect-host'),
    caFile: values.get('--ca-file'),
  };
}

if (basename(process.argv[1] ?? '') === 'protocol-smoke.js') {
  main().catch(error => {
    console.error(error instanceof Error ? error.message : error);
    process.exitCode = 1;
  });
}
